aten/src/ATen/cuda/detail/CUDAHooks.cpp

```diff
@@ -53,9 +53,65 @@ std::unique_ptr<THCState, void (*)(THCState*)> CUDAHooks::initCUDA() const {
       });
 }
 
-std::unique_ptr<Generator> CUDAHooks::initCUDAGenerator(
-    Context* context) const {
-  return std::unique_ptr<Generator>(new CUDAGenerator(context));
+// Ensures we only call cudaGetDeviceCount only once.
+static std::once_flag num_gpu_init_flag;
+
+// Total number of gpus in the system.
+static int64_t num_gpus;
+
+// Ensures default_gens_cuda is initialized once.
+static std::deque<std::once_flag> cuda_gens_init_flag;
+
+// Default, global CUDA generators, one per GPU.
+static std::vector<std::shared_ptr<CUDAGenerator>> default_gens_cuda;
+
+/* 
+* Populates the global variables related to CUDA generators
+* Warning: this function must only be called once!
+*/
+static void initCUDAGenVector(){
+  num_gpus = c10::cuda::device_count();
+  cuda_gens_init_flag.resize(num_gpus);
+  default_gens_cuda.resize(num_gpus);
+}
+
+/**
+ * PyTorch maintains a collection of default generators that get
+ * initialized once. The purpose of these default generators is to
+ * maintain a global running state of the pseudo random number generation,
+ * when a user does not explicitly mention any generator.
+ * getDefaultCUDAGenerator gets the default generator for a particular
+ * cuda device.
+ */
+CUDAGenerator* CUDAHooks::getDefaultCUDAGenerator(DeviceIndex device_index) const {
+  std::call_once(num_gpu_init_flag, initCUDAGenVector);
+  DeviceIndex idx = device_index;
+  if (idx == -1) {
+    idx = c10::cuda::current_device();
+  }
+  TORCH_CHECK(idx >= 0 && idx < num_gpus);
+  std::call_once(cuda_gens_init_flag[idx], [&] {
+    default_gens_cuda[idx] = std::make_shared<CUDAGenerator>(idx);
+    default_gens_cuda[idx]->set_current_seed(at::detail::getNonDeterministicRandom(true));
+    default_gens_cuda[idx]->set_philox_offset_per_thread(0);
+  });
+  return default_gens_cuda[idx].get();
+}
+
+/**
+ * Utility to create a CUDAGenerator. Returns a shared_ptr
+ */
+std::shared_ptr<CUDAGenerator> CUDAHooks::createCUDAGenerator(DeviceIndex device_index) const {
+  std::call_once(num_gpu_init_flag, initCUDAGenVector);
+  DeviceIndex idx = device_index;
+  if (idx == -1) {
+    idx = c10::cuda::current_device();
+  }
+  TORCH_CHECK(idx >= 0 && idx < num_gpus, "The device_index is invalid.");
+  auto gen = std::make_shared<CUDAGenerator>(idx);
+  gen->set_current_seed(default_rng_seed_val);
```

Comment:
Wow! Hard coded seed for every fresh generator? Isn't it easy to shoot yourself in the foot this way? (Shouldn't it be seeded based off of the state of the current generator? Or is this bug-for-bug compatible with the old code?)

Reply:
The old code:
```
CUDAGenerator::CUDAGenerator(Context * context_)
  : Generator(Device(DeviceType::CUDA)), context(context_)
{
}
```
I suppose did the seeding based off of the state of the current generator. But I think it's doing something weird. Every new instance of CUDAGenerator currently would point to the THCGenerators initialized by THCRandom_init. Since, we don't have usages of `torch.Generator(device='cuda')` in the current code base, we don't see any issue with this. I think `torch.Generator(device='cuda')` should give a fresh CUDAGenerator instance with a known state, and hence initializing with `default_rng_seed_val` and philox offset of 0. If somebody wanted to fork off of the current default CUDAGenerator state, they can create a CUDAGenerator and can use the clone() method?